### ISSUE TITLE
Problem in return graph query

### ISSUE THREAD

**Issue by hoptical** [opencypher/cypher-for-apache-spark#671]
Hi,
I'm trying to return graph in a simple query as below:
```
val myGraph = graph.cypher(
    s"""MATCH (n:Disease {name:"acne"})<-[r]-(m:Disease) 
    |CONSTRUCT
    |MERGE ALL n, m, r 
    |RETURN GRAPH""".stripMargin).graph
```
but I'm facing this error:
```
org.opencypher.v9_0.util.SyntaxException: Invalid input 'n': expected whitespace, comment, '=', node labels, MapLiteral, a parameter, a relationship pattern, ON, FROM GRAPH, CONSTRUCT, LOAD CSV, START, MATCH, UNWIND, MERGE, CREATE UNIQUE, CREATE, SET, DELETE, REMOVE, FOREACH, WITH, CALL, RETURN, UNION, ';' or end of input (line 3, column 11 (offset: 69))
	at org.opencypher.v9_0.parser.Base$$anonfun$parseOrThrow$1.apply(Base.scala:139)
	at org.opencypher.v9_0.parser.Base$$anonfun$parseOrThrow$1.apply(Base.scala:127)
	at scala.collection.immutable.List.map(List.scala:273)
	at org.opencypher.v9_0.parser.Base$class.parseOrThrow(Base.scala:127)
	at org.opencypher.v9_0.parser.CypherParser.parseOrThrow(CypherParser.scala:22)
	at org.opencypher.v9_0.parser.CypherParser.parse(CypherParser.scala:29)
	at org.opencypher.v9_0.frontend.phases.Parsing$.process(Parsing.scala:26)
	at org.opencypher.v9_0.frontend.phases.Parsing$.process(Parsing.scala:22)
	at org.opencypher.v9_0.frontend.phases.Phase$$anonfun$transform$1.apply(Phase.scala:35)
	at org.opencypher.v9_0.frontend.helpers.package$$anonfun$closing$1.apply(package.scala:23)
	at org.opencypher.v9_0.frontend.helpers.package$$anonfun$closing$1.apply(package.scala:23)
	at org.opencypher.v9_0.frontend.helpers.package$.using(package.scala:31)
	at org.opencypher.v9_0.frontend.helpers.package$.closing(package.scala:23)
	at org.opencypher.v9_0.frontend.phases.Phase$class.transform(Phase.scala:34)
	at org.opencypher.v9_0.frontend.phases.Parsing$.transform(Parsing.scala:22)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.v9_0.frontend.phases.PipeLine.transform(Transformer.scala:59)
	at org.opencypher.okapi.ir.impl.parse.CypherParser$class.process(CypherParser.scala:59)
	at org.opencypher.okapi.ir.impl.parse.CypherParser$.process(CypherParser.scala:39)
	at org.opencypher.okapi.relational.api.graph.RelationalCypherSession$$anonfun$3.apply(RelationalCypherSession.scala:138)
	at org.opencypher.okapi.relational.api.graph.RelationalCypherSession$$anonfun$3.apply(RelationalCypherSession.scala:138)
	at org.opencypher.okapi.relational.api.graph.RelationalCypherSession.time(RelationalCypherSession.scala:106)
	at org.opencypher.okapi.relational.api.graph.RelationalCypherSession.cypherOnGraph(RelationalCypherSession.scala:138)
	at org.opencypher.okapi.relational.api.graph.RelationalCypherSession.cypherOnGraph(RelationalCypherSession.scala:63)
	at org.opencypher.okapi.api.graph.PropertyGraph$class.cypher(PropertyGraph.scala:104)
	at org.opencypher.okapi.relational.impl.graph.ScanGraph.cypher(ScanGraph.scala:43)
	at Main$.delayedEndpoint$Main$1(Main.scala:58)
	at Main$delayedInit$body.apply(Main.scala:20)
	at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.App$class.main(App.scala:76)
	at Main$.main(Main.scala:20)
	at Main.main(Main.scala)
```
I couldn't find a documentation to check correctness of my query. 
What part of my query is wrong?
Thanks

CAPS version: 0.2.0
Spark Version: 2.2
Scala Version: 2.11.8

**Comment by DarthMax** [comment]
Hej Hamed, 

you seem to be using a slightly wrong syntax. CAPS currently does not use the `MERGE` keyword to express graph construction. 

To achieve what you want you could run the following query

```scala
val myGraph = graph.cypher(
    s"""MATCH (n:Disease {name:"acne"})<-[r]-(m:Disease) 
    |CONSTRUCT
    |  CREATE (m)-[COPY OF r]->(n) 
    |RETURN GRAPH""".stripMargin).graph
```
The `CREATE` keyword takes a pattern. Every node variable used in that pattern, that is already bound, is merged into the new graph, i.e. there will be exactly one instance of that node in the resulting graph. It is currently not allowed to merge relationships (because we would have to make sure that they are merged with their respective start and end nodes). Instead you can use the `COPY OF` keyword to create a copy of any bound relationship (copies type and properties).

We are currently working on a more comprehensive documentation. Meanwhile you can find lots of examples describing the CAPS specific Cypher syntax and our API at the examples folder: https://github.com/opencypher/cypher-for-apache-spark/tree/master/spark-cypher-examples

**Comment by hoptical** [comment]
Hey @DarthMax 

Thanks for your great answer. That worked like a charm!
But it is too slow! my neo4j database has only 4 nodes and relationships but it's still too slow!
Is there any way to cache intermediate constructed graph? like this command:
```
myGraph.cache
or
myGraph.persist
```
and about the documentation, I've read all of the examples but of course somethings like that `COPY OF' stuff cannot be found on examples.
So sorry if I ask questions on issues and thanks again for rapid answering of your team.

Best regards

**Comment by DarthMax** [comment]
Yes so one problem you seem to be encountering here is that the Sparks planning time is quite long at times - this happens irregardless of the data size. You can use `PrintRelationalPlan.set()` before running any query to print the relational plans that are compiled from the Cypher query. 

Yes, you can simply try to run `graph.cache`, this will cache all the dataframes that make up the graph.

We are working on a real documentation which will be published soon :)

Best, Max

**Comment by hoptical** [comment]
@DarthMax, 
Thanks again.

I just tried to myGraph.cache command, but it says that there is no `cache` method for `CypherResultGraph` and it is not available for `PropertyGraph` too and not either for `CypherResult` type!
Is that because of using CAPS v2.0? Should I use v1.6?

Best, Hamed

**Comment by DarthMax** [comment]
@hoptical oh I'm sorry I thought that we had introduced `#cache` on `PropertyGraph` already.
You can do it like this then:

```scala
import org.opencypher.spark.impl.CAPSConverters._
val myGraph: CypherResult#Graph = graph.cypher(
      s"""MATCH (n:Disease {name:"acne"})<-[r]-(m:Disease) 
         |CONSTRUCT
         |  CREATE (m)-[COPY OF r]->(n) 
         |RETURN GRAPH""".stripMargin).graph
myGraph.asCaps.cache()
```

**Comment by hoptical** [comment]
@DarthMax,
I tried your recommended caching method and I cached the `myGraph` cypher result as you said like this:
```scala
import org.opencypher.spark.impl.CAPSConverters._
// making graph
val myGraph: CypherResult#Graph = graph.cypher(
      s"""MATCH (n:Disease {name:"acne"})<-[r]-(m:Disease) 
         |CONSTRUCT
         |  CREATE (m)-[COPY OF r]->(n) 
         |RETURN GRAPH""".stripMargin).graph
// caching the graph
myGraph.asCaps.cache()

// get result as a dataframe from the graph
val result = myGraph.cypher(
  s"""MATCH (n)-[r]-(m)
       |RETURN r
    """.stripMargin)
val result_frame = result.records.asDataFrame
result_frame.show
```
Then I measured the run-time. It took about 15s.
Then I removed the caching line .i.e `myGraph.asCaps.cache()`, and suprisingly it's run-time was the same as before. 
This means the caching did not make any speed up!
And with 15 seconds run-time for a simple database and a simple query, cannot do any analytical procedure.
I checked up spark planning time and could not infer anything from it.

Thanks.
Best Regards

**Comment by DarthMax** [comment]
@hoptical as you are only using the graph once you will not see any speed up using `cache`, as the graph you are querying in your second query is only computed at that point in time. 

Regarding your performance issues, in which environment are you running your program Spark local mode, on a cluster, (from within Zeppelin/ spark shell)?

As I mentioned earlier, running this job on a graph with only four nodes will cause the program to spend most of the time setting up the spark environment/ connection to the cluster and then planning the execution.

**Comment by hoptical** [comment]
@DarthMax 
I tried below code on a database with about 2000 nodes and 3000 relationships and I'm using Spark in Standalone mode with 8 cores and 6Gb memory:
```scala
// making graph
  val myGraph = graph.cypher(
    s"""MATCH (n:Disease {name:"acne"})-[r]-(m:Disease)
       |MATCH (o:Sign {name:"infection"})
       |CONSTRUCT
       |  CREATE (m)-[COPY OF r]->(o)
       |RETURN GRAPH""".stripMargin).graph

// caching the graph
  myGraph.asCaps.cache()

// get result1 as a dataframe from the graph
  val result = myGraph.cypher(
    s"""MATCH (n)-[r]-(m)
       |RETURN r
    """.stripMargin)
  val result_frame = result.records.asDataFrame
  time {result_frame.show}

// get result2 as a dataframe from the graph
  val result2 = myGraph.cypher(
    s"""MATCH (n:Sign {name:"infection"})
       |RETURN n.name
    """.stripMargin)
  val result_frame2 = result2.records.asDataFrame
  time {result_frame2.show}

// get result3 as a dataframe from the graph
  val result3 = myGraph.cypher(
    s"""MATCH (m:Disease)
       |RETURN m.name
    """.stripMargin)
  val result_frame3 = result3.records.asDataFrame
  time {result_frame3.show}
```
Above code uses a `time` function which is used to measure block run-time.
I got these run-time results:
result1: 68s
result2: 7s
result3: 7s
I run code without caching and the result was almost the same as above. I checked the job timings in Spark UI too. It says the same. Meaning that caching is not affecting second and third query after even the first query is run and the graph is created. So the caching should affect second and third result but it didn't.

**Comment by Mats-SX** [comment]
@hoptical The aforementioned documentation is [available here](https://neo4j.com/docs/morpheus-user-guide/preview/). Note that this is documentation for Neo4j Morpheus, the commercial product that Neo4j is offering around CAPS. I did notice that `COPY OF` is missing from it; we'll add it in soon.

I wonder, how is `graph` instantiated in your code above? Where does the data live on disk originally? In a Neo4j database?

Also, if you were to run `time {result_frame.show}` twice in a row, does it take 68 seconds for both runs?


Generally, CAPS isn't optimised for latency, meaning small data sizes are likely to underperform when compared to one's intuitive expectation, and especially when compared to a database system such as Neo4j. It would be very interesting however to see how this scales as the data size increases -- what if your graph has 200k nodes instead of 2k?

**Comment by hoptical** [comment]
@Mats-SX 
I tested above code .i.e `time {result_frame.show}` thrice on a row on a neo4j dataset with 200K nodes and 200K relationships. But dataset was simpler than my own in the manner of nodes and relationships properties.

After the test I got these metrics:
35s for first, 30 for second and 35 for the third one.
As I said this dataset had more simpler properties than the 2k one. That's why the overall run-time is smaller than previous one.
